src/main/java/frc/robot/GenericPID/Testing/ArtificialMotor.java

```diff
@@ -0,0 +1,190 @@
+    //Make sure all operations are done in the dt range AFTER the controleffect is calculated.
+    //graph of how the pid time should update
+    //t .          .          .
+    //   |________| |________|
+    //    motor i0   motor i1    (you handle)
+    //  di0  di1        di2      (internal) 
+    //   i0        i1       i2   (internal)
+    //   p0        p1       p2   (internal)
+    
+    private static final boolean debug = Debug.debug_ArtificialMotor;
+    
+    
+    
+    private final double mass; //SI, kg
+    private final double kdrag; //SI, unitless
+    private final double kkdrag; //SI, units?
+    private final double force; //SI, N //TODO: should make a max accel too, this is max force? idk, this makes sense as is kinda, unless constraints are needed for external parts
+    //or make it so this is max force and you can use less, control it yourself? irdk, tbf, this is just a simulation
+    
+    //the artificial motor is simulated with tangential physics, not rotational
+    //So anytime the motor has a force applied, kdrag * v is subtracted, and 
+    //Acceleration is calculated by mass
+    
+    private ApproximateIntegral velocity = new ApproximateIntegral(0, 0); //SI, integrated; m/s
+    private ApproximateIntegral position = new ApproximateIntegral(0, 0); //SI, integrated; m
+    private double t;
+    
+    public static void main(String[] args) {
+        test();
+    }
+    public static void test() {
+        final int n = 6;
+        ArtificialMotor[] m = new ArtificialMotor[n];
+        Color[] c = new Color[]{Color.BLUE, Color.RED, Color.GREEN, Color.BLUE.darker(), Color.RED.darker(), Color.GREEN.darker()};
+        for (int i = 0; i < n/2; i++) {
+            m[i] = new ArtificialMotor(1, i * 0.2, 0, 0.5);
+        }
+        for (int i = n/2; i < n; i++) {
+            m[i] = new ArtificialMotor(1, 0, (i-n/2) * 0.2, 0.5);
+        }
+        Graph g = new Graph(new Graph.GraphConfig());
+        g.init(1000,1000, "Motor to 1 m/s");
+        
+        for(int i = 0; i < n; i++) {
+            g.addPlot(c[i]);
+            double dt = 0.01;
+            while (m[i].t() < 10) {
+                m[i].go(dt);
+                g.addPoint(m[i].t(), m[i].p(), i);
+            }
+            // while (m[i].t() < 10) {
+                //     m[i].direct(1, dt);
+            //     g.addPoint(m[i].t(), m[i].p(), i);
+            // }
+            // while (m[i].t() < 15) {
+            //     m[i].slow(dt);
+            //     g.addPoint(m[i].t(), m[i].p(), i);
+            // }
+        }
+    }
+
+    public double v() {
+        return velocity.val();
+    }
+    public double p() {
+        return position.val();
+    }
+    public double t() {
+        return t;
+    }
+        
+    public void go(double dt) {
+        double fnet = force - kdrag * velocity.val() - kkdrag * velocity.val() * Math.abs(velocity.val());
```

Comment:
just as a note, the linear and quadratic drag constants should never both be nonzero

Reply:
true, just a simulation, also needs to be able to handle both. Thoughts: how does the real motor have drag, and how could the PID controller be tweaked to optimize for this?